### ISSUE TITLE
State management without boilerplate by introducing new language features

### ISSUE THREAD

**Issue by vsmenon** [dart-lang/language#450]
@jascodes commented on [Sat Jul 13 2019](https://github.com/dart-lang/sdk/issues/37525)

Setting state should be easy and without boiler plate. Check out awesome Svelte 3 feature so fundamental should be built in at compiler level.

Check out https://svelte.dev/examples#reactive-assignments

Svelte 3 is awesome at using simple variables as state which can be used in UI. They do it by using reactive assignments. In order to be complaint with javascript they use `$:` label to mark assignment as reactive(i.e setstate). In dart we can use `<= ` or something else you fancy.

Thx

**Comment by miyoyo** [comment]
This isn't really relevant to Dart at all, as svelte is just a compiler.
This is achievable with a code analyzer and compiler already, dart has no issues with labels (beyond linting that they're unused) or even labels declared multiple times.

**Comment by jascodes** [comment]
Hey Guys

Svelte author used `$:` in order to be javascript spec complaint. Otherwise more preferable way is assignment operator(`<=`). He mension that in his Svelte 3 launch talk. 

So when we want to do non-state update assignment that should look like below
```
x = 10;
```
but when we want to update state, I purpose following
```
x <= 10;
```
instead of 
```
setState( () {
	x = 10;
  }
)
```

Thx
@vsmenon

**Comment by lrhn** [comment]
The `setState` function is not part of the Dart language, it's just a normal function declared by the Flutter framework. As such, we can't add syntax specific to that function.

You can already get the effect here by overriding the `x` setter to:
```dart
int _x;
int get x => _x;
void set x(int value) {
  setState(() {
    this._x = x;
  });
}
```
Then any time you do `x = 10`, it is implicitly wrapped in a `setState` call.

I'm assuming the Flutter people did not recommend that approach because they actually want you to batch up state changes so that `x = 10; y = 15;` does not send two state-change events. By forcing you to write the `setState` function call yourself, it becomes very *explicit* where important state changes happen.

**Comment by jascodes** [comment]
I agree with all that, I thought you guys were trying to opimise for flutter. Thats why i came up with suggestion.

Thx

**Comment by lrhn** [comment]
The `setState` function is not unique in what it does: It wraps a piece of code with a pre- and post-behavior.
I'm not sure exactly what, but it probably sets a bit on entry, and clears it again on exit, and then does something more at the end.

Those kinds of scopes are interesting in other situations, like resource allocation or adding logging around an operation.

The current syntax support is limited to `try`/`finally`, which doesn't help you do the pre/post opertions.
That's why `setState` takes a function as argument, it wants to wrap user code in `try`/`finally`, and the only way to move user code into a new dynamic scope is as a function.

So, some features which could make this *nicer* could be more verbost function arguments, like:
```dart
setState { x = 42; }
```
where the `setState { x = 42; }` is short for passing a function argument as `setState(() { x = 42; });`.
(I wonder if we can get away with omitting a trailing semicolon here!)

Another, harder to implement, option is call-by-name, where you pass code blocks directly, not as functions. It would allow the code blocks to contain `break` or `return` for the current function. That's aslo why it's harder, but it would allow proper user-written control structures, say `forever { code block }`, which is short for passing the `{ code block }` code by name to the `forever` function, which just contains ` for(;;) codeBlockArgument;`.

There have been requests for RAII-like patterns (Resource Allocation Is Initialization, a C++ term for using stack-allocated objects with destructors to control resource disposal). The current pattern is:
```dart
var resource = allocate();
try {
  ... use resource ..
} finally {
  resource.dispose();
}
```
which is verbose and doesn't scale. Functions like:
```dart
void control(Resource resurce, void Function(Resource) action) {
  try { action(resource); } finally ( resurce.dispose(); }
}
```
have not been nice enough to use: 
```dart
control(allocateResource(), (resource) {
  ... use resource...
});
```
Maybe a more control-structure-like syntax would make it more palatable. Maybe together with #443, it could be:
```dart
control(var resource = allocateResource()) {
   ... use resource ...
}
```
